### PR TITLE
Docker fixes

### PR DESCRIPTION
**Pull request by defcronyke**
Add new dev/ Makefile targets:
    
    build-dev
            install dependencies if needed,
            and rebuild docker images
    
    dev-ext
            run with an external Docker network,
            it's needed if you have a VPN client
            or something else on your machine,
            OpenVPN for example.
    
    build-dev-ext
            build images as above, but for external
            Docker network instead
    
    See new file `dev/docker-compose-ext-net.yml`.
    It's the same as the other docker-compose file,
    but with a docker network config added at the
    top.
    
    Add documentation to README.md about these new
    Makefile targets.

### CODE REVIEW

File: dev/Makefile [hunk 1]
```diff
@@ -1,6 +1,48 @@
...
+build-dev:
+ifeq (,$(wildcard /usr/bin/yarn))
+	cd ../web; \
+	sudo apt-get update && sudo apt-get install -y yarn || \
+	apt-get update && sudo apt-get install -y yarn || \
+	sudo pacman --needed --noconfirm -S yarn || \
+	pacman --needed --noconfirm -S yarn; \
+	yarn install && \
+	yarn build; \
+	cd ../dev
```
Comment (ebirger) [inline dev/Makefile:14]: maybe we can have a 'build-web-local' target as this seems unexpected for a docker based environment?

File: dev/Makefile [hunk 2]
```diff
@@ -1,6 +1,48 @@
...
+endif
+	docker-compose -f docker-compose.yml build
+
+# Use this target instead of `dev` to support VPN 
+# configs (OpenVPN or other).
+dev-ext:
```
Comment (ebirger) [inline dev/Makefile:20]: i don't really see a drawback in always using an external network - could be a better practice anyway. so maybe we can just have the regular dev target do this.

OTOH, i'd rather avoid duplicating the docker-compose (which had already been changed :))

**Review by ebirger — CHANGES_REQUESTED** [review]
Cool addition. thanks!

**Comment by defcronyke** [comment]
Okay yes I agree! External network is needed for my system. I'm not sure if other people would prefer it, but you're probably right about that. Better to not have two docker-compose files, especially since they're both almost the same.

The local dep installs are needed to happen anyway as far as I know, because you're using a host folder `web/` mounted into the Docker volume and exposed into the container. Is that what you want to do? Do you know a better way to avoid installing the deps on the host in that way in that case? Maybe I just don't know the best way to be doing that, so if you have a better idea then that's good too.

Do you want me to fix my PR, or will you be okay to do these things instead? Either way is fine with me.

**Comment by ebirger** [comment]
i can fix those.

maybe as a quick fix you can alter this PR to include the external net and i'll merge that part.

re the web deps i prefer to figure out some way to avoid touching the host - i'm new to yarn but i'd imagine this could be supported - probably by installing the dependencies in a folder inside the container and pointing to them. i'll sort it out.

**Comment by defcronyke** [comment]
Sure, I'll send updated PR soon. What I would usually do in this case if not wanting to touch the host, is to use a regular Docker volume for `web/` instead of mapping the host's `web/` folder to it, and during container build, I'd do the yarn dependencies installation, and then in Dockerfile I'd copy the `node_modules/` folder, as well as `public/`, `src/`, and anything else that's needed, into the volume.

If you do it that way, the files are still on the host, but in some Docker overlay folder, and the management of them is controlled by Docker properly so the host doesn't need to be concerned with those.

You should probably keep the build-dev (or similar) Makefile target too by the way, because sometimes docker-compose needs to be told to "force-rebuild" the images. It won't always rebuild when you try to start the containers when there's been changes, so this makes it more ergonomic IMO.

**Comment by ebirger** [comment]
That sounds great. i'll try that.

Totally agree on the build-dev target. please keep it then.

**Comment by defcronyke** [comment]
One more Docker idea (sorry for if you don't want all these tips haha):

If you want the best of both ways for the volume thing. You can keep your volume mapping to the host folder just as you have it already, but then you can make a second volume for only `node_modules/` folder, as a normal Docker overlay-style volume.

So in this case what that would achieve, is you would be able to hot-patch all the web files nicely still without needing to rebuild the Docker container, unless you add any new yarn dependencies. If you add new dependencies, you need to rebuild the container though (but in my first idea I mentioned above, you'd need to rebuild whenever you change any file, which makes dev environment not so useful).

**Comment by ebirger** [comment]
Yes, that was my inclination too - moving the `node_modules`  outside of the mapped part. i think that's the best of both worlds. just need to figure out how to make yarn yield to that approach :).

**Comment by defcronyke** [comment]
Don't try to change path of node_modules, it won't go so well. Instead it's best to mount the ./web volume/folder from host, and mount node_modules right inside that path in its normal spot. The `node_modules/` folder will only be visible inside the container while it's running in this case, but its contents will persist. So then when your container starts, you tell it to run `yarn install` first, and if needed, any deps will be installed in that Docker overlay location, and they'll come back next time you start the container without needing to be pulled from the internet every time.

**Comment by ebirger** [comment]
I pushed branch dev-web-deps-volume with this suggestion which seems to do the trick - my host stays clean (although the empty node_modules folder is created but i suppose that's reasonable).

**Pull request closed without merging**